### PR TITLE
Refactor extension points

### PR DESCRIPTION
**Pull request by dblock**
### Description

On top of #2. 

- Move `Extension` to the SDK root, matching OpenSearch Java code.
- Moved `api/Extension` and `api/ActionExtension`. Now, an extension implements a collection of extension points via multiple inheritance.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dblock** [comment]
@dbwiddis Take a look? This matches closer what we have in the Java code.

File: src/opensearch_sdk_py/api/extension.py [hunk 1]
```diff
@@ -9,14 +9,8 @@
...
 
 class Extension(ABC):
-    @property
-    @abstractmethod
-    def implemented_interfaces(self) -> list[tuple]:
-        """
-        Implementer should return a list of tuples containing the implemented interface (such as
-        Extension, ActionExtension, etc.) and the implementing class.
-        """
+    pass
```
Comment (dbwiddis) [inline src/opensearch_sdk_py/api/extension.py:16]: I see two `extension.py` classes, this one and one in the parent directory.   The parent one inherits from this one which seems somehow intentional, but ActionExtension (in this directory) also inherits from it.  Clearly there's some magic in how we're doing implemented extensions, but package structure isn't making it very clear.  Some comments on what's going on may be helpful.

File: tests/actions/internal/test_discovery_extension_request_handler.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+# The OpenSearch Contributors require contributions made to
+# this file be licensed under the Apache-2.0 license or a
+# compatible open source license.
+#
+
+import unittest
+
+
+class TestDiscoveryExtensionsRequestHandler(unittest.TestCase):
+    pass
```
Comment (dbwiddis) [inline tests/actions/internal/test_discovery_extension_request_handler.py:14]: I'm curious why the new test classes here that don't seem to do anything?

**Review by dbwiddis — COMMENTED** [review]
Generally looks good to be, but scratching my head at the extra Extension class and empty test classes.  What am I missing?

File: tests/actions/internal/test_discovery_extension_request_handler.py (hunk 2 above)
Comment (dblock) [reply tests/actions/internal/test_discovery_extension_request_handler.py:14]: Placeholders, we should fill them out.

File: src/opensearch_sdk_py/api/extension.py (hunk 1 above)
Comment (dblock) [reply src/opensearch_sdk_py/api/extension.py]: I was trying to have a base interface for classes like `ActionExtension`, but this is actually unnecessary. Removed.

**Review by dbwiddis — APPROVED** [review]

**Pull request merged**